> 🧹 Nitpick comments (1)
> > test/drenv/commands.py (1)> `300-314`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _💤 Low value_
> > > **Ruff B904 hint is a false positive here — design intentionally avoids `from e`.**
> > > Static analysis flags the missing `raise ... from err`/`from None`. However, using `raise ... from e` here would explicitly set `__cause__ = e` on the raised `PipelineError`, which contradicts the intent of `with_exception()` (which sets `self.__cause__ = None` while stashing the original exception in `.cause`). Per Python's exception model, a plain `raise` (without `from`) never touches `__cause__`, so the manual `None` assignment inside `with_exception()` correctly persists — confirmed by `test_pipeline_error_missing_executable` asserting `__cause__ is None`. This matches the identical pattern already used at Lines 218 and 253 for `run()`/`watch()`.
> > > One optional improvement: `__context__` will still be implicitly set to the `OSError`, so the default traceback printer may still show "During handling of the above exception..." noise. Using `raise ... from None` (instead of bare `raise`) would additionally set `__suppress_context__ = True` without affecting `.cause`, cleaning up the printed traceback while preserving current test behavior.

@coderabbit: Setting `__cause__` to any value (including `None`) also sets `__suppress_context__ = True`, so Python's traceback printer already suppresses the "During handling of the above exception..." message. Adding `from None` would be redundant.

Example:

```
>>> commands.run("it works")
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    commands.run("it works")
    ~~~~~~~~~~~~^^^^^^^^^^^^
  File "/Users/nir/src/ramen/test/drenv/commands.py", line 136, in run
    raise Error(args, f"Could not execute: {e}").with_exception(e)
  File "/Users/nir/src/ramen/test/drenv/commands.py", line 127, in run
    p = subprocess.Popen(
        args,
    ...<4 lines>...
        cwd=cwd,
    )
  File "/opt/homebrew/Cellar/python@3.14/3.14.6/Frameworks/Python.framework/Versions/3.14/lib/python3.14/subprocess.py", line 1039, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        pass_fds, cwd, env,
                        ^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                        gid, gids, uid, umask,
                        ^^^^^^^^^^^^^^^^^^^^^^
                        start_new_session, process_group)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.14/3.14.6/Frameworks/Python.framework/Versions/3.14/lib/python3.14/subprocess.py", line 1990, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
drenv.commands.Error: command failed:
  command:
  - it works
  error: 'Could not execute: [Errno 2] No such file or directory: ''it works'''
```

--- a/test/drenv/commands.py
+++ b/test/drenv/commands.py
@@ -45,7 +45,7 @@ class Error(Exception):
 
     def with_exception(self, exc):
         """
-        Return a new error preserving the traceback from another excpetion.
+        Return a new error preserving the traceback from another exception.
         """
         self.__cause__ = None
         self.cause = exc

--- a/test/drenv/commands.py
+++ b/test/drenv/commands.py
@@ -33,6 +33,10 @@ _READ_BUF = 32 * 1024
 
 class Error(Exception):
 
+    # Set by with_exception() to the original exception that caused
+    # this error (e.g. FileNotFoundError for missing executables).
+    cause = None
+
     def __init__(self, command, error, exitcode=None, output=None):
         self.command = command
         self.error = error
@@ -44,6 +48,7 @@ class Error(Exception):
         Return a new error preserving the traceback from another excpetion.
         """
         self.__cause__ = None
+        self.cause = exc
         return self.with_traceback(exc.__traceback__)
 
     def __str__(self):

--- a/test/drenv/commands_test.py
+++ b/test/drenv/commands_test.py
@@ -253,6 +253,7 @@ def test_watch_error_missing_executable():
         e.value.error,
     )
     assert e.value.__cause__ is None
+    assert isinstance(e.value.cause, FileNotFoundError)
 
 
 def test_watch_error_empty():
@@ -401,6 +402,7 @@ def test_run_error_missing_executable():
         e.value.error,
     )
     assert e.value.__cause__ is None
+    assert isinstance(e.value.cause, FileNotFoundError)
 
 
 def test_run_error_empty():
@@ -411,6 +413,7 @@ def test_run_error_empty():
     assert e.value.exitcode == 1
     assert e.value.output == ""
     assert e.value.error == ""
+    assert e.value.cause is None
 
 
 def test_run_error():
@@ -561,6 +564,7 @@ def test_pipeline_error_missing_executable():
             ["true"],
         )
     assert e.value.__cause__ is None
+    assert isinstance(e.value.cause, FileNotFoundError)
 
 
 def test_pipeline_second_fail():

